dll/win32/shell32/CShellLink.cpp

```diff
@@ -1766,8 +1767,40 @@ HRESULT STDMETHODCALLTYPE CShellLink::GetIconLocation(UINT uFlags, PWSTR pszIcon
 
 HRESULT STDMETHODCALLTYPE CShellLink::Extract(PCWSTR pszFile, UINT nIconIndex, HICON *phiconLarge, HICON *phiconSmall, UINT nIconSize)
 {
-    UNIMPLEMENTED;
-    return E_FAIL;
+    // do like ExtractIconEx but add link arrow
+    SHFILEINFOW info;
+    HIMAGELIST himl;
+
+    if (phiconLarge)
+        ZeroMemory(phiconLarge, sizeof(HICON) * nIconSize);
+    if (phiconSmall)
+        ZeroMemory(phiconSmall, sizeof(HICON) * nIconSize);
+
+    if (phiconLarge)
+    {
+        for (UINT i = 0; i < nIconSize; ++i)
```

Comment:
Can you explain to me what these loops are for?
According to [IExtractIcon::Extract method (MSDN)](https://msdn.microsoft.com/en-us/library/windows/desktop/bb761850(v=vs.85).aspx) documentation, which is the interface that CShellLink::Extract implements, the nIconSize is just a variable whose LOWORD and HIWORD specify the sizes in pixels of the large and small icons, respectively.
It doesn't sound as it has something to do with the number of "items" inside phiconLarge/Small.
Could you confirm?

Reply:
See also this apitest: https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/apitests/shell32/SHCreateFileExtractIconW.cpp;hb=e0f5ded66ac732fef4dcb96e54fc1d6cdb622714#l184
That Extract() call should work.

Reply:
I was wrong in interpretation of nIconSize. I didn't check MSDN about it. I will fix CShellLink::Extract.

Reply:
Done.